packages/amplify-category-function/src/provider-utils/awscloudformation/utils/lambda-layer-cloudformation-template.ts

```diff
@@ -0,0 +1,130 @@
+import { Fn, DeletionPolicy, Refs } from 'cloudform';
+import _ from 'lodash';
+import Lambda from 'cloudform-types/types/lambda';
+import { Permission, LayerParameters, getLayerMetadataFactory, LayerMetadata } from './layerParams';
+
+function generateLayerCfnObjBase() {
+  const cfnObj = {
+    AWSTemplateFormatVersion: '2010-09-09',
+    Description: 'Lambda Layer resource stack creation using Amplify CLI',
+    Parameters: {
+      layerVersion: {
+        Type: 'String',
+        Default: '1',
+      },
+      env: {
+        Type: 'String',
+      },
+    },
+    Resources: {},
+    Conditions: {
+      HasEnvironmentParameter: Fn.Not(Fn.Equals(Fn.Ref('env'), 'NONE')),
+    },
+  };
+  return cfnObj;
+}
+
+/**
+ * generates CFN for Layer and Layer permissions when updating layerVersion
+ */
+export function generateLayerCfnObj(context, parameters: LayerParameters) {
+  const layerData = getLayerMetadataFactory(context)(parameters.layerName);
+  const outputObj = {
+    Outputs: {
+      Arn: {
+        Value: Fn.Ref('LambdaLayer'),
+      },
+      Region: { Value: Refs.Region },
+    },
+  };
+  let cfnObj = { ...generateLayerCfnObjBase(), ...outputObj };
+  const POLICY_RETAIN = DeletionPolicy.Retain;
+  const latestVersion = layerData.getLatestVersion();
+  const layer = new Lambda.LayerVersion({
+    CompatibleRuntimes: parameters.runtimes.map(runtime => runtime.cloudTemplateValue),
+    Content: {
+      S3Bucket: Fn.Ref('deploymentBucketName'),
+      S3Key: Fn.Ref('s3Key'),
+    },
+    Description: `Lambda Layer version ${latestVersion}`,
+    LayerName: parameters.layerName,
+    LicenseInfo: 'MIT',
+  });
+  layer.deletionPolicy(POLICY_RETAIN);
+  _.assign(layer, { UpdateReplacePolicy: POLICY_RETAIN });
+
+  cfnObj.Resources['LambdaLayer'] = layer;
+  Object.entries(parameters.layerVersionMap).forEach(([key]) => {
```

Comment:
Separation of concerns: Ask all the questions, then create the permissions. Much clearer to understand and troubleshoot.

Reply:
the assignLayerPermissions function doesn't ask any questions, just transforms the inputs into valid CFN